capsules/src/spi.rs

```diff
@@ -255,3 +289,287 @@ impl<'a, S: SpiMasterDevice> SpiMasterClient for Spi<'a, S> {
+        }
+    }
+
+    pub fn config_buffers(&mut self, read: &'static mut [u8], write: &'static mut [u8]) {
+        let len = cmp::min(read.len(), write.len());
+        self.kernel_len.set(len);
+        self.kernel_read.replace(read);
+        self.kernel_write.replace(write);
+    }
+
+    // Assumes checks for busy/etc. already done
+    // Updates app.index to be index + length of op
+    fn do_next_read_write(&self, app: &mut SlaveApp) {
+        let start = app.index;
+        let len = cmp::min(app.len - start, self.kernel_len.get());
+        let end = start + len;
+        app.index = end;
+
+        self.kernel_write.map(|kwbuf| {
+            app.app_write
+                .as_mut()
+                .map(|src| for (i, c) in src.as_ref()[start..end].iter().enumerate() {
+                    kwbuf[i] = *c;
+                });
+        });
+        self.spi_slave.read_write_bytes(self.kernel_write.take(), self.kernel_read.take(), len);
+    }
+}
+
+impl<'a, S: SpiSlaveDevice> Driver for SpiSlave<'a, S> {
+    fn allow(&self, _appid: AppId, allow_num: usize, slice: AppSlice<Shared, u8>) -> ReturnCode {
+        // 0: Provides an app_read buffer
+        // 1: Provides an app_write buffer
+        match allow_num {
+            0 => {
+                let appc = match self.app.take() {
+                    None => {
+                        SlaveApp {
+                            callback: None,
+                            selected_callback: None,
+                            app_read: Some(slice),
+                            app_write: None,
+                            len: 0,
+                            index: 0,
+                        }
+                    }
+                    Some(mut appc) => {
+                        appc.app_read = Some(slice);
+                        appc
+                    }
+                };
+                self.app.replace(appc);
+                ReturnCode::SUCCESS
+            }
+            1 => {
+                let appc = match self.app.take() {
+                    None => {
+                        SlaveApp {
+                            callback: None,
+                            selected_callback: None,
+                            app_read: None,
+                            app_write: Some(slice),
+                            len: 0,
+                            index: 0,
+                        }
+                    }
+                    Some(mut appc) => {
+                        appc.app_write = Some(slice);
+                        appc
+                    }
+                };
+                self.app.replace(appc);
+                ReturnCode::SUCCESS
+            }
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
+
+    #[inline(never)]
+    fn subscribe(&self, subscribe_num: usize, callback: Callback) -> ReturnCode {
+        match subscribe_num {
+            0 /* read_write */ => {
+                let appc = match self.app.take() {
```

Comment:
Because we're using the `SlaveApp::default()` now, this can be simplified to something like:

```rust
self.app.map(|app| app.callback = callback);
```

This goes for `allow` too.